Refs #28643 -- Added Reverse database function.

pope1ni opened the pull request:
Ticket [#28643](https://code.djangoproject.com/ticket/28643).

felixxm commented:
@pope1ni Thanks for this PR :+1:. `REVERSE` on Oracle doesn't work properly with unicode strings, so it is useless in our case :disappointed: moreover it s undocumented and IMO it is not save to use it.  There are some ways to emulate `REVERSE` e.g. with `UTL_RAW.REVERSE` or `REPLACE` but they're quite complicated.

pope1ni commented:
@felixxm I was aware of the undocumented nature of `REVERSE` and that it didn't work for multi-byte strings. I thought we had some precedence for this as per footnote 6 that you added in the [ticket description](https://code.djangoproject.com/ticket/28643#comment:description) unless I misunderstood.

I did a bit of [research and play](https://rextester.com/FEVLR81667) and found a couple of possible solutions that work for multi-byte strings. The `listagg(s) within group (order by n desc)` approach has the disadvantage that it looks difficult to use it in conjunction with other functions, e.g. I cannot use `LENGTH` around the subquery. On the other hand, the `xmlquery(...).getstringval()` approach doesn't have this issue...

I've knocked up a change to see how using `XMLQUERY` works out as it seems most robust...

pope1ni commented:
buildbot, test on oracle.

pope1ni commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
@pope1ni I'm really impressed by your research :rocket: :male_detective: Thanks! Using `XMLQUERY` is a part of secret wisdom :smile: I will try to go through some edge cases to check if we don't miss anything.

orf commented on django/db/models/functions/text.py:
```diff
@@ -185,6 +185,21 @@ def __init__(self, expression, text, replacement=Value(''), **extra):
+    lookup_name = 'reverse'
+
+    def as_oracle(self, compiler, connection, **extra_context):
+        return super().as_sql(
```
Should we add a brief comment here explaining why we are using `XMLQUERY`/the problems with the standard `reverse` function?

pope1ni replied:
Sure, I can add a brief comment.

pope1ni replied:
Done.

pope1ni commented:
So this seems to work well. I have added some tests using `Reverse()` in combination with other functions to ensure that works as expected. Also of note is that passing `NULL` works as expected, yielding `NULL` :guitar: (Have added a test for that too... We added lots of `NULL` tests for the new math functions, but it looks as though we need to go back through some of the other database functions too.)

pope1ni commented:
buildbot, test on oracle.

orf commented:
Out of interest, how does `string-to-codepoints` work with unicode characters that are composed of [multiple codepoints](https://en.wikipedia.org/wiki/Combining_character)? I guess with combining characters `a+b+c` would be reversed to `c+b+a` and would hopefully work the same, but I wonder if there are any unicode strings where it would not?

pope1ni commented:
> Out of interest, how does `string-to-codepoints` work with unicode characters that are composed of [multiple codepoints](https://en.wikipedia.org/wiki/Combining_character)?

It looks like there are some things that won't work sanely: https://rextester.com/FVDM46442
But at this point it is all trading one thing off against another... :man_shrugging:

orf commented:
The `ORA-19152` error you got there could cause trouble - does this not mean the `xmlquery` would work fine until an emoji is added to some text, at which point queries would explode at runtime?

Other than that the reversals look OK

pope1ni commented:
So the combining diacritical marks look wrong as they end up out of order when reversed and this is also seen in PostgreSQL:

```psql
postgres=# select reverse('→x̊y̽z̻←');
 reverse 
---------
 ←̻z̽ẙx→
(1 row)
```

As for the emojis, they do not work with the `XMLQUERY` approach. They reverse fine, however, when using the `LISTAGG` approach. After a bit more digging, it looks like the `LISTAGG` approach is the way to go after all: https://rextester.com/LNDM28784

pope1ni commented:
buildbot, test on oracle.

pope1ni commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

timgraham commented on tests/db_functions/text/test_reverse.py:
```diff
@@ -0,0 +1,45 @@
+        self.assertQuerysetEqual(
+            authors,
+            [
+                ('John Smith', 'htimS nhoJ'),
```
On MySQL and PostgreSQL, Élena and John sort in the opposite order. I think we can leave off `order_by()` and ignore ordering?

felixxm replied:
Agreed :+1:
